### ISSUE TITLE
`lat` and `long` in Annotations

### ISSUE THREAD

**Issue by bhilburn** [gnuradio/SigMF#73]
Right now, `latitude` and `longitude` are in the `core` namespace for Annotations segments. This was put in *before* we had figured out the continuously variable method for annotating geolocation (per #55).

I think this opens the question, do we still want `lat` and `long` in the `core` namespace as a field in one of the segments? And, if so, does it belong in Annotations?

We *know* that many people will be using the new multiple-recordings approach to record geolocation for a recording. The downside, if that is the only way to define geolocation, is that we create a lot of overhead if you have a stationary receiver that you want to provide the coordinates for. The upside, though, is that there is no risk of confusion regarding what to do if both a Recording of geolocation and `lat` & `long` tags exist.

Here is my suggestion: In the `global` namespace, we add a `core:geolocation` field. This field can either be a formatted lat/long coordinate, in which case it's static for the recording, or it uses whatever method for linking to another recording that we land on in #55, in which case it's obvious that it's volatile.

Thoughts?

**Comment by bhilburn** [comment]
It's worth noting that if we go with my suggested route for geolocation, we probably will want to do the same thing for antenna azimuth.

**Comment by kpreid** [comment]
Approaching the problem from a different angle: have a way to write an annotation which explicitly is equivalent to a low-sample-rate secondary recording (specifically, two equal values at the endpoints of the annotation's time interval).

Thus, for each property of a recording that might want to vary over time, it can be represented using either a "link to another recording" or an "inlined constant recording".

**Comment by mbr0wn** [comment]
What @kpreid says is good for genericity. For the very-low-rate recording (aka just a single value), I would strongly advise that we can simply write something human-readable into the metadata file. Many people won't be both moving their receivers *and* tracking GPS locations while they do so. Same goes for antenna-steering (simply because it's a non-trivial effort to accurately know azimuth etc. of the antenna at all times). For those cases, which will be the majority of cases, we don't want to add overhead, and simply allow writing some text into the metadata file.

If we use lat/long, it's obvious what to put into the fields. There is no single way of storing geolocation, though. There's probably already multiple ways of storing GPS location data, and hell, there's other things such as [what3words](https://what3words.com) out there.
What I'm suggesting is that if we store geolocation data, we don't call the field 'geolocation', but give it a name that reflects the format in which the data is stored.

**Comment by bhilburn** [comment]
Agreed with your thoughts, @mbr0wn. Keeping simple things simple is a principal goal. And yeah, deciding exactly *how* we list geolocation is going to be another discussion. I also really like @kpreid's perspective on how to think about this.

Here's an interpretation of the above, written with loose language:
```
We specify a field in the `core` namespace that belongs in `annotations`, where the value can either be an "inlined" geolocation on a URL to another recording that provides a variable location.
```
Does that accurately represent the thoughts above (in less-than-strict wording)?

So, one of the issues that this raises, that we need to address in #55, anyway, is how we call out other recordings that are attached to a 'primary' recording. For example, it might be really convenient if in the `global` object we specifically call out all of the other recordings that apply to the current file (indeed, that's why I suggested something in `global` at the top of this Issue). Here, we are saying that we'll put it in `annotations`, though, which would require applications to parse the entirety of the `annotations` object to determine if other files are relevant unless we also tag the information elsewhere, at which point we have created duplicate information.

Thoughts on this? I know this is a general topic germane to #55, but we have good context to discuss it, here, with an immediate example.

**Comment by dharasty** [comment]
I'm new to SigMF -- but very interested in this work.  So apologies in advance if this question is well-known to current participants, or already discussed thoroughly elsewhere.

Here's my question: is the concept of notating the lat/lng to solve a) notating the location of a FIXED antenna (fixed for the duration of the recording)? or b) notating the location of a MOBILE antenna (moving over the course of the recording)?

These seem like to very different use cases... and will probably require different solutions in SigMF.

**Comment by bhilburn** [comment]
@dharasty - Welcome! Very happy to have you here =) Regarding fixed vs mobile, our strategy is laid out in #55, there, and the suggestion, above, of the field that can either be a file URL or a static value reflects that method. The discussion in this specific issue references a number of others, so no worries about not having the context!

@mbr0wn @kpreid @dharasty - There is still the question of where the `geolocation` field (or whatever we call it) goes, exactly. At the very top of this issue, I proposed putting it in `global` if it was fixed. The other option is just putting it as the very first annotation at sample[0]. It seems like the latter is the simpler case, although it's arguably less intuitive.

**Comment by bhilburn** [comment]
Need to have this fixed in v0.0.2

**Comment by bhilburn** [comment]
This must be resolved for first stable release. I believe the best approach is putting it in the `global` field as either a static field or a link to another recording.

**Comment by gmabey** [comment]
Regarding names, I'm afraid that the following (although it involves correct nomenclature/acronyms) isn't well-known enough to be as human readable as we'd all like https://en.wikipedia.org/wiki/IERS_Reference_Meridian
```python
...
    "core::irm_geolocation": {
        "latitude": 12.34,   # degrees of course
        "longitude": -56.78,  # degrees of course
        "altitude": 1000  # optional, meters of course
...
```
and ... clearly the units need to be spelled out instead of assuming anything.  Maybe, even though IRM isn't a household acronym, the meaning is both obvious and unambiguous?

**Comment by jacobagilbert** [comment]
@bhilburn I agree this must be resolved. 

Unfortunately, removing this field from the annotations is a breaking change; while I agree there are now better methods for time varying information, for now I think this needs to stay. We could also add the global field if desired but IMO removal needs to wait until v2.x

**Comment by dharasty** [comment]
Anything geospatial really should be using GeoJSON these days.  Why not store location data as a GeoJSON point?

  "geometry": {
    "type": "Point",
    "coordinates": [-56.78, 12.34]
  }

FWIW, I think GeoJSON allows a third ordinate in the "coordinates" tuple to represent altitude: [-56.78, 12.34, 1000] (though I admit I don't know many GeoJSON processors that use or frankly even preserve this information).

If you don't want to cram antenna altitude info into the GeoJSON, I consider "hagl" (height above ground level) to be a standard acronym used when discussing antennas.  And is implicitly clear without the need for the extraneous "altitude-mode".  I'd try to give every opportunity for the operator (creating the SigMF) to NOT have to also have an elevation database handy; if he's recording his location (2D) and his HAGL, he should be good to go.

Also, I wouldn't say IERS or IRM is a coordinate system... but I would say WGS84 is.  (Aka: the latitude/longitude system we all know.)  But that is a superfluous part of this spec if you state "this is a GeoJSON Point", as GeoJSON specifies that WGS 84 is its standard.

**Comment by jacobagilbert** [comment]
Also related to getting this resolved for v1.0 is what this field actually represents as that is not defined in the spec. One use is the location of the emitter responsible for the annotation which probably needs to stay here. There is a migration path away from having these in the core annotations namespace with the `signal.sigmf-ext` proposed in #120, but given how new that is  (not merged still) I thing removal is a poor choice.

There is also an argument to be made for keeping this here due to how simple it is (query gpsd when you write annotation, its immediately available w/o needing to reference another dataset) though that is admittedly not the strongest argument.

**Comment by jacobagilbert** [comment]
This can be added to the `global` object in v1.0 without breaking anything as far as I can tell. Removal of the existing `annotations` fields is the part I object to. Regardless of what we use, we should probably specify a format that is compliant with GeoJSON (RFC 7946) or OGC KML v2.3 as both of these have a lot of overhead beyond what is needed for SigMF.

**Comment by dharasty** [comment]
I have no input that use of GeoJSON should be for version 2.0 or sooner.  My only input: if you allow TWO different formats, that kind of breaks the standard, as then when I read the data I have to check for location information two ways.

One minor "use case" supporting use of GeoJSON: GeoJSON is the native geospatial interface for MongoDB.  (Not sure about other document databases.)  So if ever SigMF structures are stored in MongoDB, then doing geospatial queries such as "all records in the polygon" or "all records within 20 km of this point" are automatically supported with no additional data structure.  (Cool, IMO.)

**Comment by jacobagilbert** [comment]
@daniestevez can you comment on the original intention of the `annotation` -> `core:latitude` and `core:longitude` fields?

@dharasty between the two geoJSON probably makes the most sense to me, and you bring up another good reason to use it

**Comment by daniestevez** [comment]
> @daniestevez can you comment on the original intention of the annotation -> core:latitude and core:longitude fields?

I think you intended to @ someone else, as I don't remember giving any input about this topic (and I don't see any comments from me in this issue). Or maybe my memory is too bad and I'm missing something?

**Comment by jacobagilbert** [comment]
Sorry to bother you @daniestevez - @dkozel thought you might have some insight here.  Thanks for responding!

**Comment by daniestevez** [comment]
Sorry. No bother. I just couldn't imagine where this was coming from. Thanks to Derek for thinking that my feedback would be useful here.

I've gone through the tread to pick up on what has been said and I have mainly two things to say:

1. I agree with others who have pointed out that it's important to have a lightweight manner to indicate the position of a static receiver. While it's important to support the use case of a receiver that tracks its moving location with GPS, I think that the major use case will be static receivers.

2. Perhaps just to add to the confusion about how to specify the position, I will suggest to allow for options other than latitude and longitude. Latitude and longitude and altitude are fine for a "rough" position (say something you would search in Google maps). However, their definition can be a bit confusing or ambiguous. What is the ellipsoid (perhaps WGS84)? What is the datum? Is the altitude above the ellipsoid or above the geoid? If the latter, which geoid? All this makes geodetic coordinates not so practical where you need really good precision. Here the specific use case I have in mind is interferometry, where you need to know the locations of the antennas with millimetre precision. In many fields it is common to use cartesian (X, Y, Z) coordinates, often referred to the ITRF frame. It might be tempting to try to consider other reference systems as well, but the possibilities are almost endless, so we risk ending up with a system that is too complex if we allow many options.

In summary, I think it would be good to allow for geodetic (latitude, longitude and altitude) coordinates, and try to define them carefully (and have this option with the premise that it's used when accuracy is not so important), and also cartesian ITRF coordinates when high precision is a must.

Coming from a background in GNSS, I can help clarify with topic about any concepts about reference systems and so on that my pop up, so I'll keep an eye on this thread.

**Comment by dharasty** [comment]
I disagree with @daniestevez, and I agree with @daniestevez.

I disagree that using lat-lng is imprecise (using a double for each gives you sub-nanometer precision) or ambiguous (WGS84 -- the datum required by for GeoJSON -- is well defined!).  And I disagree that planar projections are "more accurate" because they are Cartesian... in fact, they are introducing approximations because they are Cartesian!

But I agree that map projections (where X,Y,Z are construed in meters and are Cartesian) are certainly easier to use, and plenty accurate for most purposes when doing work where the points are separated by less than 100 km.  (Frankly, I tend to work in the nearest UTM grid, or just spin up my own custom transverse Mercator projection centered on my dataset.)

My conclusion from all this mumbo-jumbo?  Folks who need to do stuff with very high precision will be paying close attention to their coordinate transformations, and those for whom rounding to the nearest meter or two will do fine assuming the Earth is a sphere.

In other words, I think that it is unwise for SigMF to try to solve these geodetic and coordinate reference system (CRS) issues... because they are all already well-solved problems in the community of folks who do work with geospatial data.  For example, GeoJSON, for one, used to support a tag to allow CRSs other than WGS84 lat-lng, and they _removed_ it from the standard!

I don't think we fall on the side of "less precision" or "ambiguous data" for sticking with GeoJSON.

**Comment by daniestevez** [comment]
Hi @dharasty, I also agree with your point of view. I think it's just a matter of different perspectives.

For sure WGS84 is a well defined reference system. Perhaps the way that I said that latitude and longitude are typically used when sub-cm accuracy is not important was somewhat confusing. The problem with latitude and longitude is that it is easy to treat them carelessly and mix up altitude over the geoid with altitude over the ellipsoid, use the wrong ellipsoid when converting from cartesian coordinates, or even using the wrong reference frame / datum. In fact, besides the problem with altitude, the rest of the mix-ups only give small errors, so people do that all the time.

I feel (but it's not more than a feeling) that ITRF cartesian coordinates are less prone to those kinds of mix ups. Also, and this is just a matter of background, cartesian coordinates are the natural system when you work with GNSS and interferometry. You care about things as points in space that co-rotate with Earth, so it seems awkward to convert back and forth to geodetic coordinates by projecting things on top of a somewhat arbitrary ellipsoid.

I can't ever seem to find what is the difference between ITRF2014 and the WGS84 datum. Reference systems are complicated.

But you make a very strong point that GeoJSON seems to be quite standard. I think that it is a good idea to give a schema that is as simple as possible, since applications will need to implement all the options allowed by the schema. To be honest, I think most of the use cases will be plotting things on a map, for which WGS84 (even with the pitfalls I mentioned) is totally OK.

I was just trying to bring new ideas to the discussion, not to invalidate the idea of using only GeoJSON in sigmf.

**Comment by jacobagilbert** [comment]
Another perspective on this: the metadata fields are intended to be augmenting information. If SigMF is required for a high precision geolocation application, users could pretty easily define their own `precision_geo.sigmf-ext` or something to allow that.

**Comment by dharasty** [comment]
> I can't ever seem to find what is the difference between ITRF2014 and the WGS84 datum.
> Reference systems are complicated.

On this we 1000% agree: coordinate reference systems are surprisingly complicated!

**Comment by dharasty** [comment]
If I can be so bold as to "read the room", it sounds like we might be coalescing around use of a 2D position encoded as a GeoJSON "Point".

But that leaves the question of altitude.  GeoJSON _allows_ a third ordinate in the coordinates array to be used for altitude, but I can't say I've seen this generally supported.  Seems our options are:

1. Use a GeoJSON "Point", and use the third ordinate anyways.
2. Use a GeoJSON "Feature" instead, encoding the location as the geometry and the altitude as one of the properties.
3. Use a GeoJSON "Point", and use some other ad-hoc field in the metadata for altitude.

A brief discussion (and simple examples) of item 1 and item 2 are given here: https://bl.ocks.org/fogonwater/340afe26d6aa0ad11a06380f73eb10f2

I find all of those equally suboptimal... however, I actually slightly lean toward item 3.

And, again I propose an alternative to altitude (which is generally construed to be "above mean sea level"), and instead prefer HAGL (height above ground level), which -- as someone with a bit of broadcast engineering in my background -- I take as a much more natural notation for the position of the radiating element of the antenna.

And if you haven't seen the connection between my last two thoughts, it is this: let's define _both_  "altitude" and "hagl" as optional components of the metadata (outside of the GeoJSON structure), and allow those who create the SigMF to include one or the other or neither.

**Comment by jacobagilbert** [comment]
I would lean toward option 1 and follow the GeoJSON spec which permits an optional third ordinate, and also, outside the GeoJSON point, allow for an optional separate HAGL field.

Altitude (m above the WGS84 ellipsoid) is part of GeoJSON, and is going to be generally available via GNSS devices. Processors accepting GeoJSON formatted points should not have errors with the altitude field in, and anything reading a discrete altitude field should be easily able to pull that information from the GeoJSON directly if available.

**Comment by bhilburn** [comment]
Read through everything, and here are my major take-aways:

* We need a digestable / usable approach widely supported by other software (including mapping) and semantic parsers that works for the general case. This should be what is in the `core` namespace and part of the basic spec. GeoJSON seems to be a great option for this.
* GeoJSON does **not** meet the requirements for high-precision interferometery. This requires the use of a different approach and system which may be a barrier for all SigMF tools to implement. I like the idea of creating a canonical extension to support this.

Regarding the options shared by @dharasty, I'm also in favor of Option 1, and using what is specified / provided by GeoJSON, even if its optional, as that's most likely to give us support and interoperability across other tooling. Looking at the GeoJSON spec ((RFC7046)[https://datatracker.ietf.org/doc/html/rfc7946]), we have:

> 3.1.1.  Position
> 
>    A position is the fundamental geometry construct.  The "coordinates"
>    member of a Geometry object is composed of either:
> 
>    o  one position in the case of a Point geometry,
> 
> \<snip\>
> 
>    A position is an array of numbers.  There MUST be two or more
>    elements.  The first two elements are longitude and latitude, or
>    easting and northing, precisely in that order and using decimal
>    numbers.  Altitude or elevation MAY be included as an optional third
>    element.

My understanding is that GeoJSON implements the 'geo' URI spec, which is (RFC5870)[https://datatracker.ietf.org/doc/html/rfc5870]. From that spec:

> 3.4.2.  Component Description for WGS-84
> 
>    The <latitude>, <longitude>, and <altitude> components as specified
>    in the URI scheme syntax (Section 3.3) are to be used as follows:
> 
>    o  <latitude> MUST contain the latitude of the identified location in
>       decimal degrees in the reference system WGS-84.
> 
>    o  <longitude> MUST contain the longitude of the identified location
>       in decimal degrees in the reference system WGS-84.
> 
>    o  If present, the OPTIONAL <altitude> MUST contain the altitude of
>       the identified location in meters in the reference system WGS-84.
> 
>    If the altitude of the location is unknown, <altitude> (and the comma
>    before) MUST NOT be present in the URI.  Specifically, unknown
>    altitude MUST NOT be represented by setting <altitude> to "0" (or any
>    other arbitrary value).

Ergo, our "geo" field would be:

A GeoJSON point with required TWO elements, in order of long-lat, with an optional third element, that is altitude. If altitude is not known, it is omitted. Long, Lat, and Altitude are all as defined by WGS-84.

HAGL is useful, but not provided by these specs, and so if needed would have to be brought in by an extension.

@daniestevez, @dharasty, @jacobagilbert, @gmabey - Thoughts on my description above? Agree / disagree?

**Comment by dharasty** [comment]
@bhilburn I think you are saying this: an example of the SigMF annotation to represent the location of a signal's recording would be:
```
"core::geo": {
    "type": "Point",
    "coordinates": [-56.78, 12.34]
}
```
Longitude and latitude coordinates -- in that order -- must be given in decimal degrees.
To annotate the altitude of the receiving antenna, use a third ordinate in the "coordinates" array to represent the , in meters above the WGS84 ellipsoid.

If that's what you're saying... I'm good to go.  (You did mean "geo" field literally, no?)

**Comment by dharasty** [comment]
A few misc comments... (wanted to split these out of my "I agree" comment, above)

1. To those wonder why we need the "type": "Point" field (seems like unnecessary static text, right?)... _that is the magic sauce that makes this a GeoJSON object_.  At least -- for example -- in MongoDB, if that (or other GeoJSON-compliant "type" field is present) is there, I can just automatically do spatial queries on that field.

2. I just checked my favorite Python library for spatial processing (Shapely), and it now supports the third-ordinate reliably.  Hooray!  (This lessens the concern I discussed above about "tools don't generally support the third ordinate".)

3. The GeoJSON spec (RFC-7946) is a bit of a split personality of the interpretation of the coordinates field.  On one hand, it requires the x and y ordinates to be decimal degrees and it expressly removes a CRS (coordinate reference system) identifier that would allow one to indicate use of an other coordinate system (aka a "projected coordinate system").  On the other hand, it does state "However, where all involved parties have a prior arrangement, alternative coordinate reference systems can be used without risk of data being misinterpreted."  Does this latter clause allow the GeoJSON Point to be used for high-precision interferometery?  (It is a use case that I'm just not familiar with.)

**Comment by jacobagilbert** [comment]
@dharasty great information, thanks. @bhilburn I am good with everything you suggested, though I am not opposed to the canonical inclusion of `core:HAGL` if there is demand. This has some non-geolocation use cases also so it might be nice to include.

With the major discussion points resolved I'll get a PR in so we can move specific implementation discussion there.

All others: further discussion is still welcome!

**Comment by jacobagilbert** [comment]
@dharasty for compatibility, would you recommend we define this as a point geolocation object directly

```JSON
  "global": {
    ...
    "core:geolocation": {
         "type": "Point",
         "coordinates": [34.0787916, -107.6183682, 2120]
    }
    ...
  }
```

or as a geometry for which we only support point objects:

```JSON
  "global": {
    ...
    "core:geolocation": {
       "geometry": {
         "type": "Point",
         "coordinates": [34.0787916, -107.6183682, 2120]
       },
    }
    ...
  }
```

**Comment by daniestevez** [comment]
@bhilburn Your proposal looks good to me.

@dharasty Regarding your question:
> Does this latter clause allow the GeoJSON Point to be used for high-precision interferometery? (It is a use case that I'm just not familiar with.)

For interferometry basically we need to be able to compute the distance (actually the vector) between two antennas (perhaps thousands of km away) with sub-millimetre precision. Looking at what you quote from the GeoJSON spec, this is just a matter of all parties agreeing on a reference system (so that there is no possible confusion with reference frames that are pretty close but not exactly the same), and using enough decimals. Typically people would use a cartesian ITRF system, mainly for convenience in the calculations (computing the straight line distance is easier, for example). From what you say I guess that when it is clear from the context and all parties have agreed (so this sounds like something for a SigMF extension), then ITRF could be used in GeoJSON.

**Comment by dharasty** [comment]
> @dharasty for compatibility, would you recommend we define this as a point geolocation object directly
> 
> ```json
>   "global": {
>     ...
>     "core:geolocation": {
>          "type": "Point",
>          "coordinates": [34.0787916, -107.6183682, 2120]
>     }
>     ...
>   }
> ```

I confirmed two ways that inner "geometry" tag is NOT necessary:
1) A close reading of the GeoJSON spec
2) I checked that in my own MongoDB dataset (on server version 4.0.4), that the inner "geometry" tag is NOT necessary and I can still do the geo queries.

So: stick with your first example (which is the one I quoted).

FWIW: the "geometry" tag is used in GeoJSON only for the "Feature" object, which has a "geometry" tag for the geometry of the feature, and a "properties" tag for additional key-value pairs.

Note that GeoJSON has several geometry types; I believe it is our intent that "core:geolocation" specifically be a GeoJSON "Point".  (No other GeoJSON types allowed.)  Be sure to incorporate that requirement.

**Comment by jacobagilbert** [comment]
@dharasty - ok great, thanks for checking into that. I'll update the merge request

**Comment by dharasty** [comment]
One last plug for supporting "height above ground level"... given by example, and some requirement-like language:


**Encoding 1** _(Proposed)_
```json
  "global": {
    "core:geolocation": {
         "type": "Point",
         "coordinates": [34.07879, -107.61836]
    }
  }
```
**Encoding 2** _(Proposed)_
```json
  "global": {
    "core:geolocation": {
         "type": "Point",
         "coordinates": [34.07879, -107.61836, 2120]
    }
  }
```
**Encoding 3** _(Proposed)_
```json
  "global": {
    "core:geolocation": {
         "type": "Point",
         "coordinates": [34.07879, -107.61836]
    },
    "core:antenna_hagl": 5
  }
```
Three forms of encoding geolocation are acceptable: _(Proposed)_

1. Use Encoding 1 where the 2-D position is known, but not the antenna's altitude or height.
2. Use Encoding 2 where the 3-D position is known; use the third ordinate in the "coordinates" field for antenna's location in meters above mean sea level (AMSL).
3. Use Encoding 3 where the 2-D position is known, but the antenna's absolute altitude AMSL is not known; use the "antenna_hagl" field for the antenna's height above ground level (HAGL) in meters.

Here are my practical reasons why I still am pulling for HAGL:

- The experimenter collecting SigMF data will generally have access to high-quality 2-D location; high-quality altitude is not common in the consumer-grade GPS systems I've used.  (To my recollection, easily off by 10s of meters.)
- The antenna's height above ground level is easily measurable and recordable.  It is also generally static in the case of a re-positioned or mobile antenna.  Plus: knowing the HAGL is still required to use "Encoding 2" (as it is the offset from actual "ground level" at the antenna's location).
- Digital elevation data these days is readily available; I'd rather not put the requirement on the data collector to have access to this in the field; the data consumer can more readily lookup the elevation from a database, and add the HAGL.
- There may be minor variations in digital elevation datasets, or in the mechanisms for interpolation specific elevations between grid points.  Some absolute altitude may be several meters above ground under one dataset and interpolation scheme, but several meters below with another.
- When it comes to predicting pathloss and other features of the propagation to/from an antenna (for example, using Longley-Rice/ITM), it is the HAGL of the transmitting and receiving elements that matter (and the differential terrain elevation between them).  To me in my work, HAGL matters most.  I'd like SigMF to be able to capture what matters to me.

OK, I promise, that is my last stand.  If you can see your way to add "core:antenna_hagl" at this time, great.  If the consensus here remains at "not interest", I'll let it rest.

Thanks for letting me make my case.

**Comment by jacobagilbert** [comment]
@dharasty You make good points for certain use cases. All three of those encodings are supported in #135.

**Comment by bhilburn** [comment]
@dharasty - You have won me over, as well. I am on-board, and will approve a PR that contains `core:hagl` for the reasons you describe 🙂
